day4/lib/day4.rb

```diff
@@ -47,7 +53,148 @@ def valid?
       private
 
       def fields
-        @raw_entry.split(" ").map { |raw_field| raw_field.split(":").first }
+        @raw_entry.split(" ").map { |raw_field| raw_field.split(":") }.to_h
+      end
+    end
+  end
+
+  class Part2
+    class BirthYear
+      def initialize(year)
+        @year = year.to_i
+      end
+
+      def valid?
+        @year.between?(1920, 2002)
+      end
+    end
+
+    class IssueYear
+      def initialize(year)
+        @year = year.to_i
+      end
+
+      def valid?
+        @year.between?(2010, 2020)
+      end
+    end
+
+    class ExpirationYear
+      def initialize(year)
+        @year = year.to_i
+      end
+
+      def valid?
+        @year.between?(2020, 2030)
+      end
+    end
+
+    class Height
+      attr_accessor :height, :units
+
+      def initialize(raw_input)
+        if !raw_input.nil?
+          matcher = raw_input.match(/(?<height>^\d+)(?<units>(cm|in))/)
```

Comment:
I'm glad you got to named capture groups. It's one of my favorite RegEx features and it feels criminally underused! I might be higher on regex than others though...